### PR TITLE
own: blob view header

### PR DESCRIPTION
**Pull request by limitedmage**
Closes https://github.com/sourcegraph/sourcegraph/issues/47816

Show a bar at the top of the blob view with history and ownership information.

Clicking the ownership side opens the ownership panel.

Also closes https://github.com/sourcegraph/sourcegraph/issues/48229

## Test plan

Storybook tests added.

## Screenshots & responsiveness

image

https://user-images.githubusercontent.com/206864/221324079-92787724-1b87-41f4-9a60-9d9b0701f21b.mp4

## App preview:

- [Web](https://sg-web-jp-blobheader.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.01% (-0.29 kb) | 0.03% (+4.80 kb) | 0.04% (+5.09 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/jp/blobheader/compare-report.html;tab=live_object) for a full comparison between the commits 8a37842 and 7f05787 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/repo/blob/own/FileOwnershipPanel.tsx [hunk 1]
```diff
@@ -64,6 +68,7 @@ export const FileOwnershipPanel: React.FunctionComponent<
                             reasons={ownership.reasons.filter(reason => reason.__typename === 'CodeownersFileEntry')}
                         />
                     ) : (
+                        // TODO: Add support for teams.
```
Comment (eseliger) [inline client/web/src/repo/blob/own/FileOwnershipPanel.tsx:71]: let's create a ticket for that pre-starship

File: client/web/src/repo/blob/own/grapqlQueries.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+export const FETCH_OWNERS_AND_HISTORY = gql`
+    ${OWNER_FIELDS}
+    ${gitCommitFragment}
+
+    query FetchOwnersAndHistory($repo: ID!, $revision: String!, $currentPath: String!) {
+        node(id: $repo) {
+            ... on Repository {
+                commit(rev: $revision) {
+                    blob(path: $currentPath) {
+                        ownership(first: 2) {
```
Comment (eseliger) [inline client/web/src/repo/blob/own/grapqlQueries.ts:57]: here we fetch 2 but in the storybook we return 3, should we return just 2 so it matches the real-world behavior?

File: client/web/src/repo/blob/own/grapqlQueries.ts [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                            nodes {
+                                owner {
+                                    ...OwnerFields
+                                }
+                            }
+                            totalCount
+                        }
+                    }
+                    ancestors(first: 1, query: "", path: $currentPath) {
```
Comment (eseliger) [inline client/web/src/repo/blob/own/grapqlQueries.ts:66]: is query not nullable? 🤔 `null` and `""` might have different meanings 🤔

File: client/web/src/repo/blob/own/HistoryAndOwnBar.tsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    revision?: string
+    filePath: string
+}> = ({ repoID, revision, filePath }) => {
+    const navigate = useNavigate()
+
+    const openOwnershipPanel = useCallback(() => {
+        navigate({ hash: '#tab=ownership' })
+    }, [navigate])
+
+    const { data, loading } = useQuery<FetchOwnersAndHistoryResult, FetchOwnersAndHistoryVariables>(
```
Comment (eseliger) [inline client/web/src/repo/blob/own/HistoryAndOwnBar.tsx:29]: no error handling?

File: client/web/src/repo/blob/own/HistoryAndOwnBar.tsx [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+            {ownership && (
+                <Tooltip content="Show ownership details" placement="left">
+                    <Button className={styles.own} onClick={openOwnershipPanel}>
+                        <div className={styles.ownBranding}>
+                            <Icon svgPath={mdiAccount} aria-hidden="true" className={styles.ownIcon} /> Own
+                        </div>
+
+                        <div className={styles.ownItems}>
+                            {ownership.nodes.slice(0, 2).map((ownership, index) => (
+                                // There will only be 2 onwers max and they won't change, so
```
Comment (eseliger) [inline client/web/src/repo/blob/own/HistoryAndOwnBar.tsx:74]:
```suggestion
                                // There will only be 2 owners max and they won't change, so
```

**Review by eseliger — APPROVED** [review]
Nice work!

**Review by cbart — APPROVED** [review]

File: client/web/src/repo/blob/own/HistoryAndOwnBar.tsx (hunk 5 above)
Comment (limitedmage) [reply client/web/src/repo/blob/own/HistoryAndOwnBar.tsx:74]: I keep making this typo 🤣

File: client/web/src/repo/blob/own/FileOwnershipPanel.tsx (hunk 1 above)
Comment (limitedmage) [reply client/web/src/repo/blob/own/FileOwnershipPanel.tsx:71]: Filed #48303 for this!

File: client/web/src/repo/blob/own/grapqlQueries.ts (hunk 2 above)
Comment (limitedmage) [reply client/web/src/repo/blob/own/grapqlQueries.ts:57]: I'm returning 3 in the storybook to correctly simulate the bad pagination we have now, to make sure that even if the backend returns more than 2 we still only display 2.

File: client/web/src/repo/blob/own/HistoryAndOwnBar.tsx (hunk 4 above)
Comment (limitedmage) [reply client/web/src/repo/blob/own/HistoryAndOwnBar.tsx]: Added more explicit error handling with a nice alert display.
image

Comment (eseliger) [reply client/web/src/repo/blob/own/HistoryAndOwnBar.tsx]: nice!

File: client/web/src/repo/blob/own/grapqlQueries.ts (hunk 3 above)
Comment (limitedmage) [reply client/web/src/repo/blob/own/grapqlQueries.ts:66]: Though it does look like it's not required so I'm removing it!

**Pull request merged**